src/main/java/strategy/PurchaseOperation.java

```
@@ -0,0 +1,18 @@
+package strategy;
+
+import db.Storage;
+import model.FruitTransaction;
+
+public class PurchaseOperation implements OperationHandler {
+
+    @Override
+    public void apply(FruitTransaction transaction) {
+        String fruit = transaction.getFruit();
+        int current = Storage.getQuantity(fruit);
+        int newBalance = current - transaction.getQuantity();
```

You subtract `transaction.getQuantity()` without validating the quantity itself. The requirements state: "Quantity must be an integer and >= 0; throw on negative quantities." Ensure the `quantity` is >= 0 before applying business logic (even if earlier conversion normally enforces this, add a defensive check here to fail fast and clearly).